_posts/2021-08-07-build-a-protected-web-app-with-reactjs-and-forgerock.md

```diff
@@ -0,0 +1,462 @@
+
+```js
+if (!step) {
+  return <Loading message='Checking your session ...' />;
+} else if (step?.callbacks?.length) {
+  return (
+    <form className='cstm_form'>
+      {step.callbacks.map(mapCallbacksToComponents)}
+      <button className='btn btn-primary w-100' type='submit'>
+        Sign In
+      </button>
+    </form>
+  );
+} else {
+  return <Error message={renderStep.payload.message} />;
+}
+```
+
+Refresh the page, and you should now have a dynamic form that reacts to the callbacks returned from our initial call to ForgeRock.
+
+## Handling the login form submission
+
+Of course, a form that can’t submit anything is not very usable, so the next step will be handling the submission. First, let’s edit the current form element, `<form className="cstm_form">`, and add an `onSubmit` handler with a simple function.
+
+```js
+<form
+  className="cstm_form"
+  onSubmit={(event) => {
+    event.preventDefault();
+    const nextStep = FRAuth.next(step);
+    setStep(nextStep);
+  }}
+>
+```
+
+You may be asking, “How did the user’s input values get added to the `step` object?” Let’s take a look at the component for rendering the username input. Open up the `Text` component: `components/journey/text.js`. Notice how special methods are being used on the callback object. These are provided as convenience methods for getting and setting data.
+
+```js
+export default function Text({ callback, inputName }) {
+  const [state] = useContext(AppContext);
+  const existingValue = callback.getInputValue();
+
+  function setValue(event) {
+    callback.setInputValue(event.target.value);
+  }
+
+  return (
+    <div className={`cstm_form-floating form-floating mb-3`}>
+      <input
+        className={`cstm_form-control form-control ${validationClass} bg-transparent ${state.theme.textClass} ${state.theme.borderClass}`}
+        defaultValue={existingValue}
+        id={inputName}
+        name={inputName}
+        onChange={setValue}
+        placeholder={textInputLabel}
+      />
+      <label htmlFor={inputName}>{textInputLabel}</label>
+    </div>
+  );
+}
+```
+
+The two important items to focus on is the `callback.getInputValue()` and the `callback.setInputValue()`. The `getInputValue` is used to display any existing value that may be provided by ForgeRock, and the `setInputValue` is used to set the user’s input on the callback while they are typing (`onChange`). The reason this works, even though we are not communicating this value to the `Form` component, is due to us passing the `callback` to the components by “reference” (not by “value”). So, any mutation of the callback object within the `Text` (or `Password`) component is also contained within the `step` object in `Form`.
+
+NOTE: You may be thinking, “That’s not very idiomatic React! Shared, mutable state is bad.” And, yes, you are correct, but we are taking advantage of this to keep everything simple (and this guide from being too long), so I hope you can excuse the pattern.
```

Comment:
is it clear that. `setInputValue` is mutative? Is it? Or am I missing a clear mutation elsewhere?